### ISSUE TITLE
Add ATSupported checks to ARIA rules

### ISSUE THREAD

**Issue by WilcoFiers** [dequelabs/axe-core#918]
One of the difficulties with the ARIA rules is that it is unclear if an ARIA attribute / role wasn't supported, or if they made had a typo. To solve this we should include all available ARIA attributes and add checks that flag unsupported properties as violations. This has the added benefit of making it easy to ignore unsupported properties by simply disabling these checks.

**Comment by WilcoFiers** [comment]
On second thought, perhaps a separate rule instead... Thoughts?

**Comment by marcysutton** [comment]
This would help to differentiate between these two scenarios, which are currently pretty cryptic:
- attribute refers to an invalid or non-existing ID attribute
- the attribute is completely invalid

**Comment by marcysutton** [comment]
> include all available ARIA attributes 
Like adding everything from the ARIA spec to the lookupTable? We can't rely on aria-query completely due to its reliance on Sets, IIRC?

> add checks that flag unsupported properties as violations
It seems like the actual marking of support should live in the lookupTable, and the checks simply look up those flags and report them to the user. Disabling the new checks would put axe back to how it operates now.

Is there anything else you want to add @WilcoFiers?

**Comment by WilcoFiers** [comment]
I don't have it all worked out... I do think we should use aria-query. We can't use it directly, but there's probably a way that we can use it to build an object structure that we can safely include into axe.

As for where the support data should be. Probably somewhere as a data structure in `commons.aria`. I think it should probably be its own structure, rather than weave it into lookuptable, so that it's easier to configure.

If you're not sure how to proceed, maybe a quick design doc would help.

**Comment by WilcoFiers** [comment]
Some more thoughts. I think we should turn whatever we do into one, or possibly multiple JSON files, rather than have them live in JS code. All this is static data, lets store it that way.

**Comment by WilcoFiers** [comment]
I'm liking this plan a lot :-). One tweak to make to it. for `axe.commons.support()`, use it as a setter. It could work something like this:

```js
axe.commons.support({
  roles: { figure: false }
})
```

**Comment by marcysutton** [comment]
Here's something I found that could be related: the [`aria-allowed-attr-matches`](https://github.com/dequelabs/axe-core/blob/develop/lib/rules/aria-allowed-attr-matches.js) file requires some kind of role to find out whether an attribute is allowed, which is probably too granular for the new `unsupported-attr` check. I think it would be good to look at all ARIA attributes present on the node for support regardless of role, because people might put them wherever.

The matches file also duplicates some of the same work as the [`aria-allowed-attr`](https://github.com/dequelabs/axe-core/blob/develop/lib/checks/aria/allowed-attr.js) check evaluate function, so that code already exists elsewhere to filter nodes out. Do you agree we should change the matches file? It would be good to confirm before I start messing with the tests, since it changes the rule design a bit. https://github.com/dequelabs/axe-core/blob/develop/test/rule-matches/aria-allowed-attr-matches.js
Thanks! :)

**Comment by dylanb** [comment]
closing after consultation
